### ISSUE TITLE
/usr/share/drcsim/setup.sh takes a long time

### ISSUE THREAD

**Issue by osrf-migration** [osrf/drcsim#274]
**Original report ([archived issue](https://osrf-migration.github.io/drcsim-gh-pages/#!/osrf/drcsim/issues/274)) by Jesper Smith (Bitbucket: [jespersmith](https://bitbucket.org/%7B61d9fe7d-e554-408c-99fb-faefa9c96114%7D/)).**

----------------------------------------
The setup.sh script uses find to find the correct drcsim directory, but this takes a long time (especially when it doesn't have an SSD, like our gazebo computers). 

Especially when adding setup.sh to your ~/.bashrc starting a new terminal or logging in to a computer takes many seconds.

**Comment by osrf-migration** [comment]
**Original comment by Jesper Smith (Bitbucket: [jespersmith](https://bitbucket.org/%7B61d9fe7d-e554-408c-99fb-faefa9c96114%7D/)).**

----------------------------------------

* Edited issue description* changed title from "/usr/share/drcsim/setup.sh makes login take a long time" to "/usr/share/drcsim/setup.sh takes a long time"

**Comment by osrf-migration** [comment]
**Original comment by John Hsu (Bitbucket: [hsu](https://bitbucket.org/%7B0a186eae-abf0-4514-a951-23db5eccc286%7D/), GitHub: [hsu](https://github.com/hsu)).**

----------------------------------------
what does

~~~
rospack profile
~~~

say?

**Comment by osrf-migration** [comment]
**Original comment by John Hsu (Bitbucket: [hsu](https://bitbucket.org/%7B0a186eae-abf0-4514-a951-23db5eccc286%7D/), GitHub: [hsu](https://github.com/hsu)).**

----------------------------------------

* set assignee_account_id to "557058:f3968cd3-4910-4384-8349-482a6c7889ec"
* set assignee to "hsu (Bitbucket: [hsu](https://bitbucket.org/hsu), GitHub: [hsu](https://github.com/hsu))"

**Comment by osrf-migration** [comment]
**Original comment by Jesper Smith (Bitbucket: [jespersmith](https://bitbucket.org/%7B61d9fe7d-e554-408c-99fb-faefa9c96114%7D/)).**

----------------------------------------

```
#!
rospack profile
Full tree crawl took 0.035247 seconds.
Directories marked with (*) contain no manifest.  You may
want to delete these directories.
To get just of list of directories without manifests,
re-run the profile with --zombie-only
-------------------------------------------------------------
0.019400   /opt/ros/groovy/share
0.005565   /opt/ros/groovy/stacks
0.003885 * /opt/ros/groovy/share/common-lisp
0.003816 * /opt/ros/groovy/share/common-lisp/ros
0.001861 * /opt/ros/groovy/share/doc
0.001801 * /opt/ros/groovy/share/doc/pcl-1.6
0.001776 * /opt/ros/groovy/share/doc/pcl-1.6/tutorials
0.001654 * /opt/ros/groovy/share/doc/pcl-1.6/tutorials/sources
0.000652   /usr/share/drcsim-2.5/ros
0.000525   /opt/ros/groovy/stacks/arm_navigation
0.000517   /opt/ros/groovy/stacks/navigation
0.000376 * /opt/ros/groovy/share/common-lisp/ros/control_msgs
0.000376   /opt/ros/groovy/stacks/pr2_controllers
0.000364   /opt/ros/groovy/stacks/arm_navigation_experimental
0.000359 * /usr/share/drcsim-2.5/gazebo_models/environments/vrc_firehose_long
0.000351 * /usr/share/drcsim-2.5/gazebo_models/environments/gate
0.000336 * /usr/share/drcsim-2.5/gazebo_models/environments/vrc_valve
0.000279   /opt/ros/groovy/share/sandia-hand-5.1/ros
0.000272 * /usr/share/drcsim-2.5/gazebo_models/environments/pavement_plane
0.000271 * /usr/share/drcsim-2.5/gazebo_models/environments/platform_1

```

**Comment by osrf-migration** [comment]
**Original comment by Brian Gerkey (Bitbucket: [Brian Gerkey](https://bitbucket.org/%7B3d8b5763-9dad-4be7-85c2-a7ab6cd49173%7D/), GitHub: [gerkey](https://github.com/gerkey)).**

----------------------------------------
If `rospack` is returning that quickly, then the problem is with the [find call](https://github.com/osrf/drcsim/blob/207da4536c151c406df38c64f605331b9d139098/cmake/setup-unversioned.sh.in#L7) that identifies the latest installed version.  I suppose that, on a system with a lot installed in `/usr/share`, that command could be slow.

You can get rid of half of the delay by sourcing the versioned setup file instead, e.g.:

    . /usr/share/drcsim-2.6/setup.sh

However, that file internally sources `/usr/share/gazebo/setup.sh`, which has the [same find call](https://bitbucket.org/osrf/gazebo/src/e7a3fa0ce69fbc112716fe2760d3907e47409a6f/cmake/setup-unversioned.sh.in?at=default#cl-7).

I suppose that the versioned `drcsim/setup.sh` could be auto-generated to source the versioned `gazebo/setup.sh` instead of the unversioned one as it does now.  Then you could avoid the delay by source the versioned `drcsim/setup.sh`.  Does that sound reasonable?

Also, how important is this problem?

**Comment by osrf-migration** [comment]
**Original comment by Jesper Smith (Bitbucket: [jespersmith](https://bitbucket.org/%7B61d9fe7d-e554-408c-99fb-faefa9c96114%7D/)).**

----------------------------------------
Avoiding the find calls would defiantly speed up things.

 It is not a big issue,  more something to fix after the vrc( hence the trivial tag)

**Comment by osrf-migration** [comment]
**Original comment by Brian Gerkey (Bitbucket: [Brian Gerkey](https://bitbucket.org/%7B3d8b5763-9dad-4be7-85c2-a7ab6cd49173%7D/), GitHub: [gerkey](https://github.com/gerkey)).**

----------------------------------------
Ah, didn't check the priority.  OK, I'll put it on hold for now.

In the meantime, a quick workaround for you would be to manually modify `/usr/share/drcsim-2.6/setup.sh` to source `/usr/share/gazebo-1.8/setup.sh` and then start using `/usr/share/drcsim-2.6/setup.sh` in your workflow.

**Comment by osrf-migration** [comment]
**Original comment by Brian Gerkey (Bitbucket: [Brian Gerkey](https://bitbucket.org/%7B3d8b5763-9dad-4be7-85c2-a7ab6cd49173%7D/), GitHub: [gerkey](https://github.com/gerkey)).**

----------------------------------------

* changed state from "new" to "on hold"

**Comment by osrf-migration** [comment]
**Original comment by Jesper Smith (Bitbucket: [jespersmith](https://bitbucket.org/%7B61d9fe7d-e554-408c-99fb-faefa9c96114%7D/)).**

----------------------------------------

* changed state from "on hold" to "open"

Should be trivial to fix. Would a simlink from /usr/share/drcsim to the latest version be possible? That way, we don't need to change the version in our model import scripts either.

**Comment by osrf-migration** [comment]
**Original comment by Brian Gerkey (Bitbucket: [Brian Gerkey](https://bitbucket.org/%7B3d8b5763-9dad-4be7-85c2-a7ab6cd49173%7D/), GitHub: [gerkey](https://github.com/gerkey)).**

----------------------------------------
From @Progtologist in [Gazebo PR #178](https://bitbucket.org/osrf/gazebo/pull-request/178):

`find @CMAKE_INSTALL_PREFIX@/share -path @CMAKE_INSTALL_PREFIX@/share/gazebo-'*'/setup.sh | sort | tail -n 1` is very slow and causes a significant lag if inserted in .bashrc A better solution should be provided. A simple workaround is to add the argument `-type f` in find to specify not to search folder names.

**Comment by osrf-migration** [comment]
**Original comment by Brian Gerkey (Bitbucket: [Brian Gerkey](https://bitbucket.org/%7B3d8b5763-9dad-4be7-85c2-a7ab6cd49173%7D/), GitHub: [gerkey](https://github.com/gerkey)).**

----------------------------------------
Similar to [gazebo issue #784](https://bitbucket.org/osrf/gazebo/issue/784).

**Comment by osrf-migration** [comment]
**Original comment by Brian Gerkey (Bitbucket: [Brian Gerkey](https://bitbucket.org/%7B3d8b5763-9dad-4be7-85c2-a7ab6cd49173%7D/), GitHub: [gerkey](https://github.com/gerkey)).**

----------------------------------------

* changed assignee_account_id from "557058:f3968cd3-4910-4384-8349-482a6c7889ec" to "557058:2e35028e-9ad8-4d6a-a3a5-275159a304a9"
* changed assignee from "hsu (Bitbucket: [hsu](https://bitbucket.org/hsu), GitHub: [hsu](https://github.com/hsu))" to "gerkey (Bitbucket: [gerkey](https://bitbucket.org/gerkey), GitHub: [gerkey](https://github.com/gerkey))"

**Comment by osrf-migration** [comment]
**Original comment by Brian Gerkey (Bitbucket: [Brian Gerkey](https://bitbucket.org/%7B3d8b5763-9dad-4be7-85c2-a7ab6cd49173%7D/), GitHub: [gerkey](https://github.com/gerkey)).**

----------------------------------------
Adding `-type f` doesn't have a noticable effect:

~~~
$ time find /usr/share -path /usr/share/drcsim-'*'/setup.sh | sort | tail -n 1
/usr/share/drcsim-2.6/setup.sh

real  0m2.492s
user  0m0.208s
sys   0m0.576s
~~~

~~~
$ time find /usr/share -type f -path /usr/share/drcsim-'*'/setup.sh | sort | tail -n 1
/usr/share/drcsim-2.6/setup.sh

real  0m2.488s
user  0m0.200s
sys   0m0.596s
~~~

In between runs, I followed [this tip](http://stackoverflow.com/questions/9551838/how-to-purge-disk-i-o-caches-on-linux) to clear disk caches via `echo 3 | sudo tee /proc/sys/vm/drop_caches`.  Otherwise, the first run is slow and the rest of the runs are much faster, once that part of the file system has been cached.

EDIT: I get the same results doing the test directly on the search for gazebo's `setup.sh`, without going through drcsim's `setup.sh`.

**Comment by osrf-migration** [comment]
**Original comment by John Hoare (Bitbucket: [jhoare](https://bitbucket.org/%7Bf2fb0aad-9933-4b16-a710-d9fe1486032f%7D/)).**

----------------------------------------
Like @jespersmith said, Isn't installing a symbolic link to the drcsim-X.X generally the way to approach this? 

If you want to avoid that, you could avoid using find: 

```ls /usr/share/drcsim-*/setup.sh | sort | tail -n 1```

Which works much faster for me.

**Comment by osrf-migration** [comment]
**Original comment by Brian Gerkey (Bitbucket: [Brian Gerkey](https://bitbucket.org/%7B3d8b5763-9dad-4be7-85c2-a7ab6cd49173%7D/), GitHub: [gerkey](https://github.com/gerkey)).**

----------------------------------------

* changed state from "open" to "resolved"

Fixed in [pull request #361](https://osrf-migration.github.io/drcsim-gh-pages/#!/osrf/drcsim/pull-requests/361) by installing the main setup.sh twice, once to the versioned directory and once to the unversioned directory. I chose that solution over using a symlink to avoid possible cross-platform or upgrade issues.

Together with [gazebo pull request #639](https://bitbucket.org/osrf/gazebo/pull-request/639), things are much faster:

~~~
$ echo 3 | sudo tee /proc/sys/vm/drop_caches
3
$ time . install/share/drcsim/setup.sh 

real	0m0.125s
user	0m0.032s
sys	0m0.008s
~~~
